impl/src/main/java/com/okta/sdk/impl/oauth2/AccessTokenRetrieverServiceImpl.java

```diff
@@ -0,0 +1,90 @@
+
+    private ObjectMapper objectMapper = new ObjectMapper();
+
+    private ClientConfiguration clientConfiguration;
+
+    public AccessTokenRetrieverServiceImpl(ClientConfiguration clientConfiguration) {
+        Assert.notNull(clientConfiguration);
+        this.clientConfiguration = clientConfiguration;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public OAuth2AccessToken getOAuth2AccessToken()
+        throws IOException, InvalidKeySpecException, NoSuchAlgorithmException {
+        String signedJwt = createSignedJWT(clientConfiguration);
+        String scope = String.join(" ", clientConfiguration.getScopes());
+
+        MediaType mediaType = MediaType.parse("application/x-www-form-urlencoded");
+        RequestBody body = RequestBody.create(mediaType, "");
+
+        OkHttpClient client = new OkHttpClient().newBuilder().build();
+
+        Request accessTokenRequest = new Request.Builder()
+            .url(clientConfiguration.getBaseUrl() +
+                "/oauth2/v1/token" +
+                "?grant_type=client_credentials" +
+                "&client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-bearer" +
+                "&client_assertion=" + signedJwt +
+                "&scope=" + scope)
+            .method(HttpMethod.POST.name(), body)
+            .addHeader("Accept", "application/json")
+            .addHeader("Content-Type", "application/x-www-form-urlencoded")
+            .build();
+
+        Response accessTokenResponse = client.newCall(accessTokenRequest).execute();
+
+        if (accessTokenResponse != null && accessTokenResponse.body() != null) {
+            Reader reader = new InputStreamReader(accessTokenResponse.body().byteStream(), StandardCharsets.UTF_8);
+
+            if (accessTokenResponse.isSuccessful()) {
+                log.debug("Access token request was successful");
+                return objectMapper.readValue(reader, OAuth2AccessToken.class);
+            }
+            else {
+                ErrorResponse errorResponse = objectMapper.readValue(reader, ErrorResponse.class);
+                log.debug("Access token request failed with error {}", errorResponse);
+
+                if (!Strings.isEmpty(errorResponse.getError()) ||
+                    !Strings.isEmpty(errorResponse.getErrorDescription())) {
+                    throw new HttpException("Received HTTP " + accessTokenResponse.code() +
+                        " response from Authorization Server. error - " + errorResponse.getError() +
+                        ", error description - " + errorResponse.getErrorDescription());
+                } else {
+                    throw new HttpException("Received HTTP " + accessTokenResponse.code() +
+                        " response from Authorization Server");
+                }
+            }
+        }
+        else {
+            throw new HttpException("Could not retrieve access token from Authorization Server");
```

Comment:
I'm not familiar with Java exceptions, but is `HttpException` the right one here?

Reply:
I believe so :)

Reply:
Nope, its an Exception specific to this library

Reply:
Sorry @bdemers , whats your recommendation here?

Reply:
Just answering the question ;)